sway/tree/arrange.c

```diff
@@ -18,30 +18,51 @@ static void apply_horiz_layout(list_t *children, struct wlr_box *parent) {
 		return;
 	}
 
-	// Calculate total width of children
-	double total_width = 0;
+	// Count the number of new windows we are resizing, and how much space
+	// is currently occupied
+	int new_children = 0;
+	double current_width_fraction = 0;
 	for (int i = 0; i < children->length; ++i) {
 		struct sway_container *child = children->items[i];
-		if (child->width <= 0) {
-			if (children->length > 1) {
-				child->width = parent->width / (children->length - 1);
+		current_width_fraction += child->width_fraction;
+		if (child->width_fraction <= 0) {
+			new_children += 1;
+		}
+	}
+
+	// Calculate each height fraction
+	double total_width_fraction = 0;
+	for (int i = 0; i < children->length; ++i) {
+		struct sway_container *child = children->items[i];
+		if (child->width_fraction <= 0) {
+			if (current_width_fraction <= 0) {
+				child->width_fraction = 1.0;
+			} else if (children->length > new_children) {
+				child->width_fraction = current_width_fraction /
+					(children->length - new_children);
 			} else {
-				child->width = parent->width;
+				child->width_fraction = current_width_fraction;
 			}
 		}
-		container_remove_gaps(child);
-		total_width += child->width;
+		total_width_fraction += child->width_fraction;
+	}
+	// Normalize width fractions so the sum is 1.0
+	for (int i = 0; i < children->length; ++i) {
+		struct sway_container *child = children->items[i];
+		child->width_fraction /= total_width_fraction;
+
+		sway_log(SWAY_DEBUG, "Child %d with %lf width_fraction", i, child->width_fraction);
```

Comment:
style: 80 columns

Reply:
These are all stray changes from debugging stuff. I've removed them although it may make sense to add a little bit more to the log to be able to diagnose weird things that crop up in the future. Not sure how much.